flang/lib/Lower/Bridge.cpp

```diff
@@ -1481,6 +1481,22 @@ class FirConverter : public Fortran::lower::AbstractConverter {
         curEval = &*std::next(curEval->getNestedEvaluations().begin());
       }
     }
+    auto ompSections =
+        std::get_if<Fortran::parser::OpenMPSectionsConstruct>(&omp.u);
+    // If a 'sections' construct is encountered, loop over all 'section'
```

Comment:
The problem was that FIR was messing up. The farthest I could get was

```
omp.sections{
    omp.section{
        
         omp.section{
         }
   }
}
```

I could understand that this is probably happening because once you call the function `OpenMP.cpp`, you call `genFIR()` as soon as the operation is created. However, we here want a proper nested `omp.section` operations with the `genFIR()` putting it FIR into respective `omp.section`

That was the rationale behind this situation.

Reply:
The bridge should function correctly where nested constructs are recursively built. So if getNestedEvaluations is returning _any_ omp op with a Region, it should recursively fill out and build the appropriate context. It looks like the next loop may be expecting erroneously that it can flatten all high-level OpenMP constructs with scope (just ignores there is a Region possibility).

Reply:
Thanks for the comment @schweitzpgi. So if I understand correctly, you are suggesting that we can create `sectionOp` inside of `genOMP(...., OpenMPSectionsConstruct)` through observing the nested evaluations. And once we return back to `genFIR(const Fortran::parser::OpenMPConstruct &omp)` in Bridge.cpp, the last loop in this function will automatically fill in the FIR to each of the `sectionOp` separately.  Precisely this piece of code here:

https://github.com/flang-compiler/f18-llvm-project/blob/8111f26793a60cbfb11244524a45368114586167/flang/lib/Lower/Bridge.cpp#L1485

I also understand that probably this nested `sectionOp` was observed because I did not handle the `insertionPt` correctly. Although this is an assumption; I do not know the exact reason.

Reply:
Thanks @NimishMishra for the explanation. I think you can emphasize in the comment that the individual `Section(s)` inside the `Sections` construct are `SectionBlocks` and not constructs, hence the nested constructs handling don't apply here. Due to the structure of the `Sections` construct and the `SectionBlocks` inside the construct we need special handling here. 
I guess the alternative is to have a separate function for `SectionBlocks` just like the `DeclarativeConstruct` but is probably not worth it.

Reply:
@NimishMishra , the idea would be to build up a recursive construct in the PFT and here in the bridge just naturally lower it. The PFT construct might hold a "section" (or whatever) which can itself hold more sections, etc. The insertion point would have to necessarily be set to the interior of the construct being visited and reset when the bridge finishes lowering it.

To @kiranchandramohan point, I don't know if it would be worthwhile or not. Perhaps "section" here is a highly unusual case or something. But if there is a general pattern where you have OMP ops that hold an mlir::Region then it might be worth the trouble to lower the entire group of constructs generically via recursive calls.

Reply:
Thanks @kiranchandramohan and @schweitzpgi for the discussion on previous revision. Also adding @shraiysh for comment.

As suggested, I revisited the problem of trying to fit `sectionOp` in `OpenMP.cpp` rather than in `Bridge.cpp` as is the current state. I did not succeed in the same. However, I spotted a major flaw in my understanding of `pft::Evaluation` which had led to erroneous code (now fixed). I am summarising the same here to explain the current revision's state as well as to discuss if we can still fit `sectionOp` in `OpenMP.cpp`.

Previous iterations of this PR's lowering assumed something like the following:

```
sectionsOp
   - sectionOp
         - EvaluationPartConstruct (some statement for which to generate FIR)
         - EvaluationPartConstruct
          ...
   - sectionOp
         - EvaluationPartConstruct
         - EvaluationPartConstruct
         ...
```

However, I found out that `pft::Evaluation` is unaware of the `sectionOp` entirely. I did this test:

**[What the previous iterations of this PR had]** Wherein we generate one `sectionOp` for every nested evaluation of `sectionsOp`. Expected result is we get `sectionOp`s equal to the number of actual `!$omp section` enclosed.  Precisely the following code:

```
for(auto &e : curEval->getNestedEvaluations())
      //generate 1 sectionOp here
```
However, if you check the attached screenshot, it is clear that `getNestedEvaluations()` returned a list of Evaluations whose number was equal to `EvaluationPartConstruct` and not equal to the number of `!$omp section`. This means previous iterations of this PR were generating **one sectionOp for each executable statement within !$omp sections**

![what_was_wrong](https://user-images.githubusercontent.com/42909663/145585546-208c5ab0-3c0f-4e6c-9aae-cca91492f20d.png)

**[What the current version has]** Currently, I pick up a list of nested `pft::Evaluations`, extract the number of `EvaluationPartConstruct`s in each of the `!$omp section` within `!$omp sections` and generate FIR that goes with the proper `sectionOp`.

Reply:
Hi @kiranchandramohan I was working on parallel sections, and was wondering why can't we call `genFIR` from `AbstractConverter`? IMO it has a clear distribution of responsibilities, where the construct decides how its nested evaluations must be lowered. This is also similar to the `genType` and `genExpr*` which are already exposed in `AbstractConverter`. I have added a small draft patch showing what I mean [here](https://reviews.llvm.org/D121476). Please have a look at it and let me know if it sounds like a good idea - I will cherry-pick the handling of clauses and tests from this patch in that case.

Reply:
> @schweitzpgi I think `Sections` and the `section`s inside it are a kind of special constructs. So will prefer special handling here. Also we cannot do mutually recursive call since we cannot call genFIR from code in OpenMP.cpp since only the `AbstractConverter` is available there.
> 
> I like the idea of making relevant changes in the PFT. But I think this will require creating new parse-tree nodes for the `Section` Construct under `OpenMPConstruct`. I will prefer to move forward for now with special handling. Since the patch has waited a long time I am planning to merge this today.

Having the PFT create a nested structure (like other Fortran constructs) does sound like what would be best here as well.

A construct is a sufficiently large enough unit that it makes sense for the Bridge to perform the choreography itself and not pass it back and forth to "helpers" in OpenMP.cpp, which should deal with smaller, more atomic parts of lowering. The bridge should own all the context, analysis, etc.

I didn't follow the parse tree changes comment. It's possible to construct PFT data structures that are not simple pointers into the parse tree, if that is required. ?

Reply:
> I didn't follow the parse tree changes comment. It's possible to construct PFT data structures that are not simple pointers into the parse tree, if that is required. ?

Currently only Lowering of `OpenMPConstruct`s and `OpenMPDeclarativeConstruct`s are forwarded to OpenMP.cpp for further processing. If we create an `Evaluation` that is not one of these it will probably require further `genFIR` hooks in the bridge right?